src/additional_zones.rs

```diff
@@ -0,0 +1,216 @@
+            .collect::<Vec<_>>()
+            .into_par_iter()
+            .map(|(parent, mut places)| compute_voronoi(parent, &mut places, &zones, &towns))
+            .flatten()
+            .collect()
+    };
+    publish_new_cities(zones, new_cities);
+}
+
+fn get_parent<'a>(place: &Zone, zones: &'a [Zone], zones_rtree: &ZonesTree) -> Option<&'a Zone> {
+    zones_rtree
+        .fetch_zone_bbox(&place)
+        .into_iter()
+        .map(|z_idx| &zones[z_idx.index])
+        .filter(|z| z.contains_center(place))
+        .min_by_key(|z| z.zone_type)
+}
+
+fn is_place(obj: &OsmObj) -> bool {
+    match *obj {
+        OsmObj::Node(ref node) => node
+            .tags
+            .get("place")
+            .map_or(false, |v| v == "city" || v == "town" || v == "village"),
+        _ => false,
+    }
+}
+
+fn read_places(pbf_path: &str) -> Vec<Zone> {
+    let path = Path::new(&pbf_path);
+    let file = File::open(&path).expect("no pbf file");
+
+    let mut parsed_pbf = OsmPbfReader::new(file);
+    let mut zones = vec![];
+
+    for obj in parsed_pbf.par_iter().map(Result::unwrap) {
+        if !is_place(&obj) {
+            continue;
+        }
+        if let OsmObj::Node(ref node) = obj {
+            let next_index = ZoneIndex { index: zones.len() };
+            if let Some(mut zone) = Zone::from_osm_node(&node, next_index) {
+                if zone.name.is_empty() {
+                    continue;
+                }
+                zone.zone_type = Some(ZoneType::City);
+                zone.center = Some(Point::<f64>::new(node.lon(), node.lat()));
+                zone.bbox = zone.center.as_ref().map(|p| Rect {
+                    min: Coordinate {
+                        x: p.0.x - std::f64::EPSILON,
+                        y: p.0.y - std::f64::EPSILON,
+                    },
+                    max: Coordinate {
+                        x: p.0.x + std::f64::EPSILON,
+                        y: p.0.y + std::f64::EPSILON,
+                    },
+                });
+                zone.is_generated = true;
+                zones.push(zone);
+            }
+        }
+    }
+    zones
+}
+
+fn convert_to_geo(geom: GGeom) -> Option<MultiPolygon<f64>> {
+    match geom.try_into().expect("conversion to geo failed") {
+        geo::Geometry::Polygon(x) => Some(MultiPolygon(vec![x])),
+        y => {
+            if let Some(x) = y.into_multi_polygon() {
+                Some(x)
+            } else {
+                None
+            }
+        }
+    }
+}
+
+fn extrude_existing_town(zone: &mut Zone, towns: &[ZoneWithGeos]) {
+    if towns.is_empty() {
+        return;
+    }
+    if let Some(ref mut boundary) = zone.boundary {
+        let mut updates = 0;
+        let mut g_boundary = boundary.try_into().expect("failed to convert to geos");
+        for town in towns {
+            if g_boundary.intersects(&town.geos).unwrap_or_else(|_| false) {
+                if let Ok(b) = g_boundary.difference(&town.geos) {
```

Comment:
* why not using the difference directly without checking for the intersection ?
* it looks expensive to iter through all towns (potentially in the world). It could be useful to use the ZonesTree here too, to filter candidate towns to extrude.